File: ocaml/day2.ml
open Base

type input = (string * int) list

let parse_cmd s =
  let op, n = String.lsplit2_exn s ~on:' ' in
  (op, Int.of_string n)

let parse_input lines = List.map ~f:parse_cmd lines

let apply_cmd (pos, depth) = function
  | "forward", x -> (pos + x, depth)
  | "down", x -> (pos, depth + x)
  | "up", x -> (pos, depth - x)
  | cmd, _ -> failwith ("Unknown command " ^ cmd)

let part1 cmds =
  let pos, depth = List.fold_left ~init:(0, 0) ~f:apply_cmd cmds in
  pos * depth

let apply_cmd2 (pos, depth, aim) = function
  | "forward", x -> (pos + x, depth + (aim * x), aim)
  | "down", x -> (pos, depth, aim + x)
  | "up", x -> (pos, depth, aim - x)
  | cmd, _ -> failwith ("Unknown command " ^ cmd)

let part2 cmds =
  let pos, depth, _ = List.fold_left ~init:(0, 0, 0) ~f:apply_cmd2 cmds in
  pos * depth

File: haskell/day2.hs
module Day2 where

import Util
import qualified Aoc

parseCmd s = 
  let (op, n) = splitPair ' ' s in
  (op, read n)

parse = map parseCmd
  
applyCmd (pos, depth) cmd = case cmd of
  ("forward", x)    -> (pos + x, depth)
  ("down", x)       -> (pos, depth + x)
  ("up", x)         -> (pos, depth - x)
  _ -> error ("Unknown command " ++ show cmd)

part1 cmds = 
  let (pos, depth) = foldl applyCmd (0,0) cmds in
  pos * depth

applyCmd2 (pos, depth, aim) cmd = case cmd of
  ("forward", x)    -> (pos + x, depth + (aim * x), aim)
  ("down", x)       -> (pos, depth, aim + x)
  ("up", x)         -> (pos, depth, aim - x)
  _ -> error ("Unknown command " ++ show cmd)

part2 cmds =
  let (pos, depth, _) = foldl applyCmd2 (0,0,0) cmds in
  pos * depth

run = Aoc.run 2 parse part1 part2

